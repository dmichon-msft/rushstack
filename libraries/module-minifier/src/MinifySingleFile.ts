// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { minify, MinifyOptions, MinifyOutput, SimpleIdentifierMangler } from 'terser';
import type { RawSourceMap } from 'source-map';

declare module 'terser' {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  interface SourceMapOptions {
    asObject?: boolean;
  }
}

import { getIdentifier } from './MinifiedIdentifier';
import type {
  IModuleMinificationRequest,
  IModuleMinificationResult,
  IModuleMinificationSuccessResult
} from './types';

const nth_identifier: SimpleIdentifierMangler = {
  get: getIdentifier
};

/**
 * Minifies a single chunk of code. Factored out for reuse between WorkerPoolMinifier and LocalMinifier
 * @internal
 */
export async function minifySingleFileAsync(
  request: IModuleMinificationRequest,
  terserOptions: MinifyOptions,
  useDecodedMap?: boolean
): Promise<IModuleMinificationResult> {
  const { code, nameForMap, hash, externals } = request;

  try {
    const {
      format: rawFormat,
      output: rawOutput,
      mangle: originalMangle,
      ...remainingOptions
    } = terserOptions;

    const format: MinifyOptions['format'] = rawFormat || rawOutput || {};

    const mangle: MinifyOptions['mangle'] =
      originalMangle === false ? false : typeof originalMangle === 'object' ? { ...originalMangle } : {};

    const finalOptions: MinifyOptions = {
      ...remainingOptions,
      format,
      mangle
    };
    format.comments = false;

    if (mangle) {
      mangle.nth_identifier = nth_identifier;
    }

    if (mangle && externals) {
      mangle.reserved = mangle.reserved ? externals.concat(mangle.reserved) : externals;
    }

    finalOptions.sourceMap = nameForMap
      ? {
          asObject: true
        }
      : false;

    const minified: MinifyOutput = await minify(
      {
        [nameForMap || 'code']: code
      },
      finalOptions
    );

    const result: IModuleMinificationSuccessResult = {
      hash,
      error: undefined,
      code: minified.code!,
      decodedMap: useDecodedMap ? minified.decoded_map : undefined,
      map: useDecodedMap ? undefined : minified.map
    };

    return result;
  } catch (error) {
    console.error(error);
    return {
      hash,
      error: error as Error,
      code: undefined,
      decodedMap: undefined,
      map: undefined
    };
  }
}
