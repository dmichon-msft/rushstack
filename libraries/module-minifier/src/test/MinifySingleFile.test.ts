// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { minifySingleFileAsync } from '../MinifySingleFile';

describe(minifySingleFileAsync.name, () => {
  it('uses consistent identifiers for webpack vars', async () => {
    const code: string = `__MINIFY_MODULE__(function (module, __webpack_exports__, __webpack_require__) {});`;

    const minifierResult = await minifySingleFileAsync(
      {
        hash: 'foo',
        code,
        nameForMap: undefined,
        externals: undefined
      },
      {
        mangle: true
      }
    );

    expect(minifierResult).toMatchSnapshot();
  });

  it('allows performing only dead code elimination', async () => {
    const code: string = `__MINIFY_MODULE__(function (module, __webpack_exports__, __webpack_require__) {/* comment */var x;\n/* other comment */var y = 10;\nconsole.log(y);});`;

    const minifierResult = await minifySingleFileAsync(
      {
        hash: 'foo',
        code,
        nameForMap: undefined,
        externals: undefined
      },
      {
        compress: {
          defaults: false,
          dead_code: true,
          unused: true
        },
        format: {
          comments: 'all'
        },
        mangle: false
      }
    );

    expect(minifierResult).toMatchSnapshot();
  });
});
