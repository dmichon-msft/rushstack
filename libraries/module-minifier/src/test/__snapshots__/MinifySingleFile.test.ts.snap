// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minifySingleFileAsync allows performing only dead code elimination 1`] = `
Object {
  "code": "__MINIFY_MODULE__((function(module,__webpack_exports__,__webpack_require__){
/* other comment */var y=10;console.log(y)}));",
  "error": undefined,
  "hash": "foo",
  "map": undefined,
}
`;

exports[`minifySingleFileAsync uses consistent identifiers for webpack vars 1`] = `
Object {
  "code": "__MINIFY_MODULE__((function(e,t,n){}));",
  "error": undefined,
  "hash": "foo",
  "map": undefined,
}
`;
